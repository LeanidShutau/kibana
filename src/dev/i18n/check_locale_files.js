/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import path from 'path';
import JSON5 from 'json5';

import { difference, globAsync, readFileAsync } from './utils';
import { verifyJSON } from './verify_locale_json';

export async function checkFile(localePath) {
  let errorMessage = '';

  const defaultMessagesBuffer = await readFileAsync(
    path.resolve(path.dirname(localePath), 'en.json')
  );
  const defaultMessagesIds = Object.keys(JSON5.parse(defaultMessagesBuffer.toString()));

  const localeBuffer = await readFileAsync(localePath);

  let namespace;
  try {
    namespace = verifyJSON(localeBuffer.toString(), localePath);
  } catch (error) {
    throw new Error(`Error in ${path.relative(__dirname, localePath)}\n${error.message || error}`);
  }

  const translations = JSON5.parse(localeBuffer.toString());
  const translationsIds = Object.keys(translations);

  const unusedTranslations = difference(translationsIds, defaultMessagesIds);
  const missingTranslations = difference(defaultMessagesIds, translationsIds);

  if (unusedTranslations.length > 0) {
    errorMessage += `\nUnused translations in locale file ${path.relative(__dirname, localePath)}:
${unusedTranslations.join(', ')}`;
  }

  if (missingTranslations.length > 0) {
    errorMessage += `\nMissing translations in locale file ${path.relative(__dirname, localePath)}:
${missingTranslations.join(', ')}`;
  }

  if (errorMessage) {
    throw new Error(errorMessage);
  }

  return namespace;
}

export async function checkLocaleFiles(pluginsPaths) {
  const pluginsMapByLocale = new Map();

  for (const pluginPath of pluginsPaths) {
    const globOptions = {
      ignore: ['./translations/en.json', './translations/messagesCache.json'],
      cwd: path.resolve(pluginPath),
    };

    const localeEntries = await globAsync('./translations/*.json', globOptions);

    for (const entry of localeEntries) {
      const locale = path.basename(entry);

      if (pluginsMapByLocale.has(locale)) {
        pluginsMapByLocale.get(locale).push(pluginPath);
      } else {
        pluginsMapByLocale.set(locale, [pluginPath]);
      }
    }
  }

  for (const locale of pluginsMapByLocale.keys()) {
    const namespaces = [];
    for (const pluginPath of pluginsMapByLocale.get(locale)) {
      const namespace = await checkFile(path.resolve(pluginPath, 'translations', locale));
      if (namespaces.includes(namespace)) {
        throw new Error(
          `Error in ${path.relative(__dirname, pluginPath)} plugin ${locale} locale file
Locale file namespace should be unique for each plugin`
        );
      }
      namespaces.push(namespace);
    }
  }
}
