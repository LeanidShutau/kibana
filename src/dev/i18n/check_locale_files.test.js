/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import path from 'path';

import { checkFile, checkLocaleFiles } from './check_locale_files';

const testsFixturesRoot = path.resolve(__dirname, '__fixtures__', 'check_locale_files');

const pluginsPaths = [
  path.join(testsFixturesRoot, 'test_plugin_1'),
  path.join(testsFixturesRoot, 'test_plugin_2'),
  path.join(testsFixturesRoot, 'test_plugin_3'),
  path.join(testsFixturesRoot, 'test_plugin_4'),
  path.join(testsFixturesRoot, 'test_plugin_5'),
];

describe('dev/i18n/check_locale_files', () => {
  describe('checkFile', () => {
    it('returns namespace of a valid JSON file', async () => {
      const localePath1 = path.join(pluginsPaths[0], 'translations', 'valid.json');
      const localePath2 = path.join(pluginsPaths[1], 'translations', 'valid.json');

      expect(await checkFile(localePath1)).toBe('test_plugin_1');
      expect(await checkFile(localePath2)).toBe('test_plugin_2');
    });

    it('throws an error for unused id and missing id', async () => {
      const localeWithMissingMessage = path.join(pluginsPaths[2], 'translations', 'missing.json');
      const localeWithUnusedMessage = path.join(pluginsPaths[2], 'translations', 'unused.json');
      await expect(checkFile(localeWithMissingMessage)).rejects.toThrowErrorMatchingSnapshot();
      await expect(checkFile(localeWithUnusedMessage)).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('checkLocaleFiles', () => {
    it('validates locale files in multiple plugins', async () => {
      expect(await checkLocaleFiles([pluginsPaths[0], pluginsPaths[1]])).toBe(undefined);
    });

    it('throws an error for namespaces collision', async () => {
      await expect(
        checkLocaleFiles([pluginsPaths[3], pluginsPaths[4]])
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
