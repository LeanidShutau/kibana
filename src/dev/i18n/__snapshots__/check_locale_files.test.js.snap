// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dev/i18n/check_locale_files checkFile throws an error for unused id and missing id 1`] = `
"
Missing translations in locale file __fixtures__\\\\check_locale_files\\\\test_plugin_3\\\\translations\\\\missing.json:
test_plugin_3.id_2"
`;

exports[`dev/i18n/check_locale_files checkFile throws an error for unused id and missing id 2`] = `
"
Unused translations in locale file __fixtures__\\\\check_locale_files\\\\test_plugin_3\\\\translations\\\\unused.json:
test_plugin_3.id_3"
`;

exports[`dev/i18n/check_locale_files checkLocaleFiles throws an error for namespaces collision 1`] = `
"Error in __fixtures__\\\\check_locale_files\\\\test_plugin_5 plugin valid.json locale file
Locale file namespace should be unique for each plugin"
`;
